### PR TITLE
GH-41562: [C++][Parquet] Decoding: Fix num_value handling in ByteStreamSplitDecoder

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

This problem is raised from  https://github.com/apache/arrow/pull/40094 . Original bug fixed here: https://github.com/apache/arrow/pull/34140 , but this is corrupt in https://github.com/apache/arrow/pull/40094 .

### What changes are included in this PR?

Refine checking

### Are these changes tested?

* [x] Will add

### Are there any user-facing changes?

Bugfix

* GitHub Issue: #41562

### CODE REVIEW

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -3694,12 +3694,23 @@ class ByteStreamSplitDecoderBase : public DecoderImpl,
...
-    if (static_cast<int64_t>(num_values) * byte_width_ != len) {
+  void SetData(int num_values, const uint8_t* data, int len) final {
+    // Check that the data size is consistent with the number of values
+    // The spec requires that the data size is a multiple of the number of values,
+    // see: https://github.com/apache/parquet-format/pull/192 .
+    // GH-41562: passed in `num_values` may include nulls, so we need to check and
+    // adjust the number of values.
+    if (static_cast<int64_t>(num_values) * byte_width_ < len) {
       throw ParquetException("Data size (" + std::to_string(len) +
                              ") does not match number of values in BYTE_STREAM_SPLIT (" +
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:3705]:
```suggestion
                             ") is too small for the number of values in BYTE_STREAM_SPLIT (" +
```

File: cpp/src/parquet/encoding_test.cc [hunk 2]
```diff
@@ -1383,7 +1383,7 @@ class TestByteStreamSplitEncoding : public TestEncodingBase<Type> {
     encoder->PutSpaced(draws_, num_values_, valid_bits, valid_bits_offset);
     encode_buffer_ = encoder->FlushValues();
     ASSERT_EQ(encode_buffer_->size(), physical_byte_width() * (num_values_ - null_count));
-    decoder->SetData(num_values_ - null_count, encode_buffer_->data(),
+    decoder->SetData(num_values_, encode_buffer_->data(),
```
Comment (pitrou) [inline cpp/src/parquet/encoding_test.cc:1386]: Ouch. Can we also improve the docstring for `Decoder::SetData`, to make it clear that the `num_values` parameter also includes null values?

Comment (mapleFU) [reply cpp/src/parquet/encoding_test.cc:1386]: done. This part is so tricky, I may revisit it later

File: cpp/src/parquet/encoding.h [hunk 3]
```diff
@@ -255,6 +255,10 @@ class Decoder {
 
   // Sets the data for a new page. This will be called multiple times on the same
   // decoder and should reset all internal state.
+  //
+  // In current implementations, `num_values` is the `num_values` field in the
+  // data page header, which may greater than the number of values in the data
+  // buffer.
```
Comment (pitrou) [inline cpp/src/parquet/encoding.h:261]:
```suggestion
  // `num_values` comes from the data page header, and may be greater than the number of
  // physical values in the data buffer if there are some omitted (null) values.
  // `len`, on the other hand, is the size in bytes of the data buffer and
  // directly relates to the number of physical values.
```

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by pitrou — APPROVED** [review]
Thanks a lot for fixing this, @mapleFU .
CI failures seem unrelated.

**Pull request merged**